### ISSUE TITLE
CEF (Common Event Format) parser in logql

### ISSUE THREAD

**Issue by james-callahan** [grafana/loki#5648]
**Is your feature request related to a problem? Please describe.**
CEF (Common Event Format) is a logging format used by some tools. Loki and promtail should have a built in parser for it so that I can extract CEF fields as labels.
CEF messages look something like:
```
CEF:0|Palo Alto Networks|Cortex XDR|Cortex XDR 2.4|XDR Analytics|High Connection Rate|6|end=1601792870694 shost=WGHRAMG deviceFacility=None cat=Discovery externalId=98106342 request=https:\/\/iga-bh.xdr.eu.paloaltonetworks.com\/alerts\/98106342 cs1=iexplore.exe cs1Label=Initiated by cs2=\“C:\\\\Program Files (x86)\\\\Internet Explorer\\\\IEXPLORE.EXE\” SCODEF:11844 CREDAT:82946 \/prefetch:2 cs2Label=Initiator CMD cs3=Microsoft CorporationSIGNATURE_SIGNED- cs3Label=Signature cs4=iexplore.exe cs4Label=CGO name cs5=\“C:\\\\Program Files (x86)\\\\Internet Explorer\\\\IEXPLORE.EXE\” SCODEF:11844 CREDAT:82946 \/prefetch:2 cs5Label=CGO CMD cs6=Microsoft CorporationSIGNATURE_SIGNED- cs6Label=CGO Signature dst=10.12.4.37 dpt=8000 src=10.10.28.140 spt=58003 fileHash=e582676ec900249b408ab4e37976ae8c443635a7da77755daf6f896a172856a3 filePath=C:\\\\Program Files (x86)\\\\Internet Explorer\\\\iexplore.exe targetprocesssignature=NoneSIGNATURE_UNAVAILABLE- tenantname=iGA tenantCDLid=1021319191 CSPaccountname=Information & eGovernment Authority initiatorSha256=e582676ec900249b408ab4e37976ae8c443635a7da77755daf6f896a172856a3 initiatorPath=C:\\\\Program Files (x86)\\\\Internet Explorer\\\\iexplore.exe cgoSha256=e582676ec900249b408ab4e37976ae8c443635a7da77755daf6f896a172856a3 osParentName=iexplore.exe osParentCmd=\“C:\\\\Program Files (x86)\\\\Internet Explorer\\\\IEXPLORE.EXE\” SCODEF:11844 CREDAT:82946 \/prefetch:2 osParentSha256=e582676ec900249b408ab4e37976ae8c443635a7da77755daf6f896a172856a3 osParentSignature=SIGNATURE_SIGNED osParentSigner=Microsoft Corporation incident=118719 act=Detected suser=['root']
```
i.e. a pipe delimited format; where the last field is a series of key/value pairs with certain escaping rules

**Describe the solution you'd like**

Similar to LogQL parsers `json` and `logfmt`, we should have a parser `cef`

**Additional context**

[Documentation of the format](https://kc.mcafee.com/resources/sites/MCAFEE/content/live/CORP_KNOWLEDGEBASE/78000/KB78712/en_US/CEF_White_Paper_20100722.pdf)

My first use-case is to parse logs from Cortex XDR which are sent as CEF payloads over syslog (see https://docs.paloaltonetworks.com/cortex/cortex-xdr/cortex-xdr-pro-admin/logs/cortex-xdr-log-notification-formats/alert-notification-format.html)

**Comment by james-callahan** [comment]
Please label revivable?

**Comment by chaudum** [comment]
still valid

**Comment by chaudum** [comment]
The basic structure of the CEF format can be parsed using the `pattern` parser, like this: `pattern `CEF:<v>|<vendor>|<product>|<version>|<signature>|<name>|<severity>|<extension>`

Example:

```
{filename="/var/log/cef.log"} |= "cat=Discovery" | pattern `CEF:<v>|<vendor>|<product>|<version>|<signature>|<name>|<severity>|<extension>`
```

However, if you want to parse also the key value pairs of the extension it can get tricky. Even though it looks very much like `logfmt`, it's not really the case. However, you can still use regular line filters, as in the example above.

**Comment by james-callahan** [comment]
> However, if you want to parse also the key value pairs of the extension it can get tricky. Even though it looks very much like logfmt, it's not really the case. However, you can still use regular line filters, as in the example above.

At least with the logs coming from Cortex XDR, all the interesting things to alert on are in the `extension` section. We'd like to extract them to add to alerts generated from loki ruler.

**Comment by chaudum** [comment]
> > However, if you want to parse also the key value pairs of the extension it can get tricky. Even though it looks very much like logfmt, it's not really the case. However, you can still use regular line filters, as in the example above.
> 
> At least with the logs coming from Cortex XDR, all the interesting things to alert on are in the `extension` section. We'd like to extract them to add to alerts generated from loki ruler.

That's what I thought. I guess a regex guru may be able to write an expression for parsing the key value pairs, but it would be very inefficient. :(

**Comment by chaudum** [comment]
If you don't need to parse **all** key value pairs of the extension field, it may be relatively easy:


```
{filename="/var/log/cef.log"}
|= "cat=Discovery"
| pattern `CEF:<v>|<vendor>|<product>|<version>|<signature>|<name>|<severity>|<extension>`
| label_format original=`CEF:{{.v}}|{{.vendor}}|{{.product}}|{{.version}}|{{.signature}}|{{.name}}|{{.severity}}|{{.extension}}`
| line_format `{{.extension}}` 
| regexp `end=(?P<end>[^\s]+)` 
| end > 1600000000000
| line_format `{{.original}}`
```

And since you want to alert on certain things, I assume you don't even need the last `line_format` stage, because you're doing a metrics query:

```
sum by (shost) (
  count_over_time(
    {filename="/var/log/cef.log"}
    |= "cat=Discovery"
    | pattern `CEF:<v>|<vendor>|<product>|<version>|<signature>|<name>|<severity>|<extension>`
    | line_format `{{.extension}}` 
    | regexp `end=(?P<end>[^\s]+)` 
    | regexp `shost=(?P<shost>[^\s]+)` 
    | end > 1600000000000
    [$__interval]
  )
)
```

**Do not blame me on performance ;-)**

**Comment by james-callahan** [comment]
Even that doesn't help with the un-escaping process.

Though I think it does show why a dedicated `cef` parser would be useful :)

**Comment by chaudum** [comment]
> Though I think it does show why a dedicated `cef` parser would be useful :)

Agree, a dedicated parser would be more useful. If we wanna do this, we'll have some thoughts on whether integrating a parser in LogQL or in Promtail, so it can be transformed into a different format.

**Comment by james-callahan** [comment]
> we'll have some thoughts on whether integrating a parser in LogQL or in Promtail, so it can be transformed into a different format.

@chaudum any further thoughts on this?

**Comment by lg-d** [comment]
Is there any progress in making a CEF parser?

**Comment by fourstepper** [comment]
I tried searching around the internet for some CEF specification, but I soft-failed... There are some vendors that kind of explain the formatting, but some actual specification/RFC would help a lot

**Comment by dfaligertwood** [comment]
Vector have this in VRL, and have an encoder. See vectordotdev/vector#17389 - their documentation reference was [this guide](https://www.microfocus.com/documentation/arcsight/arcsight-smartconnectors-8.3/cef-implementation-standard/)

**Comment by lukasgabriel** [comment]
This would be a great addition!

**Comment by vmichalak** [comment]
I already have implemented a CEF parser in golang for one of my personal project. If someone can guide me on how to implement a parser (especially which interface i have to implement I imagine), I can purpose you an implementation.

**Comment by JStickler** [comment]
Promtail is end of life (EOL) as of March 2, 2026. No future support or updates will be provided. All future feature development will occur in [Grafana Alloy](https://github.com/grafana/alloy).
